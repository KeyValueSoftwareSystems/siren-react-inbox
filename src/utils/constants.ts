export const COLORS = {
  light: {
    primaryColor: "#F56630",
    highlightedCardColor: "#FFECE5",
    textColor: "#344054",
    neutralColor: "#FFFFFF",
    borderColor: "#D0D5DD",
    dateColor: "#34405499",
    deleteIcon: "#34405499",
    timerIcon: "#667185",
    clearAllIcon: "#667185",
    windowShadowColor: "#D0D5DD",
    iconColor: "#F7F9FC",
    notificationIcon: "#232326",
  },
  dark: {
    primaryColor: "#F56630",
    highlightedCardColor: "#2E2D30",
    textColor: "#FFFFFF",
    neutralColor: "#232326",
    borderColor: "#344054",
    dateColor: "#98A2B3",
    deleteIcon: "#D0D5DD",
    timerIcon: "#98A2B3",
    clearAllIcon: "#D0D5DD",
    windowShadowColor: "#232326",
    iconColor: "#38383D",
    notificationIcon: "#FFFFFF",
  },
};

export enum ThemeMode {
  DARK = "dark",
  LIGHT = "light",
}

export enum LogLevel {
  INFO,
  ERROR,
}

export enum BadgeType {
  NONE = "none",
  DOT = "dot",
  DEFAULT = "default",
}

export const levelLogFns = {
  [LogLevel.INFO]: console.log,
  [LogLevel.ERROR]: console.error,
};

export const defaultBadgeStyle = {
  size: 15,
  color: "#FF0000",
  textColor: "#FFFFFF",
  textSize: 10,
  linHeight: "14px",
  top: "inherit",
  left: "inherit",
};

export enum eventTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  UPDATE_NOTIFICATIONS_COUNT = "UPDATE_NOTIFICATIONS_COUNT",
  RESET_NOTIFICATIONS = "RESET_NOTIFICATIONS",
  RESET_NOTIFICATIONS_COUNT = "RESET_NOTIFICATIONS_COUNT",
}

export enum events {
  NOTIFICATION_LIST_EVENT = "NOTIFICATION_LIST_EVENT",
  NOTIFICATION_COUNT_EVENT = "NOTIFICATION_COUNT_EVENT",
}

export enum updateNotificationsTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
}

export enum sirenReducerTypes {
  SET_NOTIFICATIONS = "SET_NOTIFICATIONS",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  SET_UN_VIEWED_NOTIFICATION_COUNT = "SET_UN_VIEWED_NOTIFICATION_COUNT",
  SET_SIREN_CORE = "SET_SIREN_CORE",
}

export const LIST_EMPTY_TEXT = "No new notifications";
export const LIST_EMPTY_SUB_TEXT = "Check back later for updates and alerts.";
export const ERROR_TEXT = "Oops! Something went wrong";
export const ERROR_SUB_TEXT =
  "Could not load the notifications. Please refresh the page.";
export const DEFAULT_WINDOW_TITLE = "Notifications";
export const RETRY_BUTTON_LABEL = "Retry";
export const CLEAR_ALL_LABEL = "Clear All";
export const IN_APP_RECIPIENT_UNAUTHENTICATED = 'IN_APP_RECIPIENT_UNAUTHENTICATED';
export const TOKEN_VERIFICATION_PENDING ='TOKEN_VERIFICATION_PENDING';
export const MAXIMUM_RETRY_COUNT = 3;

export const errorMap = {
  SIREN_OBJECT_NOT_FOUND: {
    Type: "ERROR",
    Code: "SIREN_OBJECT_NOT_FOUND",
    Message: "Siren Object Not found",
  },
  MISSING_PARAMETER: {
    Type: "ERROR",
    Code: "MISSING_PARAMETER",
    Message: "Missing Parameter",
  },
};

export enum VerificationStatus {
  PENDING = "PENDING",
  SUCCESS = "SUCCESS",
  FAILED = "FAILED"
}