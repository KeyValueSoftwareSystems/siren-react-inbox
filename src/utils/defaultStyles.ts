
const defaultStyles = {
  notificationIcon: {
    size: 30,
  },
  window: {
    width: 500,
    borderRadius: "10px",
    borderWidth: "0px",
    borderStyle: "solid"
  },
  windowHeader: {
    height: 64,
    titleFontWeight: "600",
    titleSize: 20,
    closeIconSize: 20,
    titlePadding: 0,
  },
  windowContainer: {
    padding: 0,
    height: 700,
  },
  notificationCard: {
    padding: 10,
    borderWidth: 0.5,
    avatarSize: 40,
    titleFontWeight: "600",
    titleSize: 16,
    descriptionSize: 14,
    dateSize: 12,
  },
  loadMoreButton: {
    fontSize: 14,
    fontWeight: '600'
  },
  badgeStyle: {
    size: 15,
    textSize: 10,
  },
  deleteIcon:{
    size: 16
  },
  timerIcon:{
    size: 12
  },
  clearAllIcon:{
    size: 18
  }
}

export default defaultStyles;
