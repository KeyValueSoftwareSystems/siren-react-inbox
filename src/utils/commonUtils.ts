import type { RefObject } from "react";

import type {
  ActionResponse,
  NotificationDataType,
  NotificationsApiResponse,
} from "@sirenapp/js-sdk/dist/esm/types";

import {
  defaultBadgeStyle,
  eventTypes,
  LogLevel,
  ThemeMode,
} from "./constants";
import { default as DefaultStyle } from "./defaultStyles";
import { default as DefaultTheme } from "./defaultTheme";
import type {
  CustomStyle,
  DimensionValue,
  SirenStyleProps,
  ThemeProps,
} from "../types";

type FetchParams = {
  size: number;
  start?: string;
  end?: string;
  sort?: "createdAt" | "updatedAt";
};

export const generateElapsedTimeText = (timeString: string) => {
  const currentTime = new Date().getTime();
  const targetTime = new Date(timeString).getTime();
  const millisecondsDiff = currentTime - targetTime;

  const seconds = Math.floor(millisecondsDiff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  const months = Math.floor(days / 30);
  const years = Math.floor(days / 365);

  if (millisecondsDiff < 60000) return "Just now";
  else if (minutes < 60)
    return minutes === 1 ? "1 minute ago" : `${minutes} minutes ago`;
  else if (hours < 24) return hours === 1 ? "1 hour ago" : `${hours} hours ago`;
  else if (days < 30) return days === 1 ? "1 day ago" : `${days} days ago`;
  else if (months < 12)
    return months === 1 ? "1 month ago" : `${months} months ago`;
  else return years === 1 ? "1 year ago" : `${years} years ago`;
};

export const updateNotifications = (
  eventData: {
    id?: string;
    action: string;
    newNotifications?: NotificationDataType[];
    unreadCount?: number;
  },
  notifications: NotificationDataType[]
): NotificationDataType[] => {
  let updatedNotifications: NotificationDataType[] = [];

  switch (eventData.action) {
    case eventTypes.MARK_ITEM_AS_READ:
      updatedNotifications = notifications.map((item) =>
        item.id === eventData.id ? { ...item, isRead: true } : item
      );
      break;
    case eventTypes.MARK_ALL_AS_READ:
      updatedNotifications = notifications.map((item) => ({
        ...item,
        isRead: true,
      }));
      break;
    case eventTypes.DELETE_ITEM:
      updatedNotifications = notifications.filter(
        (item) => item.id !== eventData.id
      );
      break;
    case eventTypes.DELETE_ALL_ITEM:
      updatedNotifications = [];
      break;
    case eventTypes.NEW_NOTIFICATIONS: {
      const newNotifications: NotificationDataType[] =
        eventData?.newNotifications || [];

      updatedNotifications = [...newNotifications, ...notifications];
      break;
    }
    case eventTypes.RESET_NOTIFICATIONS: {
      updatedNotifications = [];
      break;
    }
    default:
      break;
  }

  return updatedNotifications;
};

export const renderAsImage = (icon: JSX.Element | string): boolean =>
  typeof icon === "undefined" || typeof icon === "string";

export const logger = {
  log: async (level: LogLevel.INFO | LogLevel.ERROR, message: string) => {
    const timestamp = new Date().toISOString();
    const levelString = LogLevel[level].toUpperCase();

    console.log(`[${timestamp}] [${levelString}] ${message}`);
  },
  error: function (error: string) {
    this.log(LogLevel.ERROR, error);
  },
  info: function (message: string) {
    this.log(LogLevel.INFO, message);
  },
};

export const applyTheme = (
  theme: ThemeProps = {},
  mode: ThemeMode = ThemeMode.DARK,
  customStyle: CustomStyle = {}
): SirenStyleProps => {
  const windowBorderRadius =
    customStyle.window?.borderRadius || DefaultStyle.window.borderRadius;

  return {
    container: {
      maxWidth: customStyle.window?.width || "100",
    },
    windowShadow: {
      boxShadow: `${
        theme.colors?.windowShadowColor || DefaultTheme[mode].window.shadowColor
      } 0px 8px 24px`,
    },
    windowTopBorder: {
      borderTopLeftRadius: windowBorderRadius,
      borderTopRightRadius: windowBorderRadius,
    },
    windowBottomBorder: {
      borderBottomLeftRadius: windowBorderRadius,
      borderBottomRightRadius: windowBorderRadius,
    },
    contentContainer: {
      backgroundColor:
        theme.windowContainer?.background ||
        theme.colors?.neutralColor ||
        DefaultTheme[mode].windowContainer.background,
      padding:
        customStyle.windowContainer?.padding ||
        DefaultStyle.windowContainer.padding,
      borderRadius: `0 0 ${
        customStyle.window?.borderRadius || DefaultStyle.window.borderRadius
      } ${
        customStyle.window?.borderRadius || DefaultStyle.window.borderRadius
      }`,
    },
    body: {
      overflow: "auto",
      height: customStyle.windowContainer?.contentHeight || "700px",
    },
    headerContainer: {
      alignItems: "center",
      display: "flex",
      margin: 0,
      lineHeight: "28px",
      backgroundColor:
        theme.windowHeader?.background ||
        theme.colors?.neutralColor ||
        DefaultTheme[mode].windowHeader.background,
      borderBottom: `${customStyle.windowHeader?.borderWidth || DefaultStyle.windowHeader.borderWidth} solid`,
      borderColor: theme.colors?.borderColor || DefaultTheme[mode].colors.borderColor,
      height:
        customStyle.windowHeader?.height || DefaultStyle.windowHeader.height,
    },
    headerTitle: {
      color:
        theme.windowHeader?.titleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].windowHeader.titleColor,
      fontSize:
        customStyle.windowHeader?.titleSize ||
        DefaultStyle.windowHeader.titleSize,
      fontWeight:
        customStyle.windowHeader?.titleFontWeight ||
        DefaultStyle.windowHeader.titleFontWeight,
      paddingLeft:
        customStyle.windowHeader?.titlePadding ||
        DefaultStyle.windowHeader.titlePadding,
    },
    headerAction: {
      color:
        theme.windowHeader?.headerActionColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].windowHeader.headerActionColor,
    },
    defaultCardContainer: {
      backgroundColor:
        theme.notificationCard?.background ||
        DefaultTheme[mode].notificationCard.background,
      padding:
        customStyle.notificationCard?.padding ||
        DefaultStyle.notificationCard.padding,
      borderBottom: `${
        customStyle.notificationCard?.borderWidth ||
        DefaultStyle.notificationCard.borderWidth
      }px solid`,
      borderColor:
        theme.notificationCard?.borderColor ||
        theme.colors?.borderColor ||
        DefaultTheme[mode].notificationCard.borderColor,
    },
    cardIconRound: {
      width:
        customStyle.notificationCard?.avatarSize ||
        DefaultStyle.notificationCard.avatarSize,
      height:
        customStyle.notificationCard?.avatarSize ||
        DefaultStyle.notificationCard.avatarSize,
      borderRadius:
        (customStyle.notificationCard?.avatarSize ||
          DefaultStyle.notificationCard.avatarSize) / 2,
      overflow: "hidden",
      backgroundColor:
        theme.colors?.borderColor || DefaultTheme[mode].colors.borderColor,
    },
    cardTitle: {
      color:
        theme.notificationCard?.titleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.titleColor,
      fontSize:
        customStyle.notificationCard?.titleSize ||
        DefaultStyle.notificationCard.titleSize,
      fontWeight:
        customStyle.notificationCard?.titleFontWeight ||
        DefaultStyle.notificationCard.titleFontWeight,
    },
    cardSubTitle: {
      color:
        theme.notificationCard?.subTitleColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.subTitleColor,
      fontSize:
        customStyle.notificationCard?.subTitleSize || DefaultStyle.notificationCard.subTitleSize,
      fontWeight:
        customStyle.notificationCard?.subTitleFontWeight ||
        DefaultStyle.notificationCard.subTitleFontWeight,
    },
    activeCardMarker: {
      backgroundColor:
        theme.colors?.highlightedCardColor ||
        DefaultTheme[mode].colors?.highlightedCardColor,
      border:
        theme.colors?.primaryColor || DefaultTheme[mode].colors?.primaryColor,
    },
    cardDescription: {
      color:
        theme.notificationCard?.descriptionColor ||
        theme.colors?.textColor ||
        DefaultTheme[mode].notificationCard.descriptionColor,
      fontSize:
        customStyle.notificationCard?.descriptionSize ||
        DefaultStyle.notificationCard.descriptionSize,
      fontWeight:
      customStyle.notificationCard?.descriptionFontWeight ||
      DefaultStyle.notificationCard.descriptionFontWeight,
    },
    dateStyle: {
      color: theme.colors?.dateColor || DefaultTheme[mode].colors.dateColor,
      fontSize:
        customStyle.notificationCard?.dateSize ||
        DefaultStyle.notificationCard.dateSize,
      lineHeight: "16px",
    },
    emptyText: {
      color: theme.colors?.textColor || DefaultTheme[mode].colors.textColor,
    },
    errorText: {
      color: theme.colors?.textColor || DefaultTheme[mode].colors.textColor,
    },
    clearIcon: {
      color: theme.colors?.clearAllIcon || DefaultTheme[mode].clearIcon.color,
      size: customStyle.clearAllIcon?.size || DefaultStyle.clearAllIcon.size,
    },
    timerIcon: {
      color: theme.colors?.timerIcon || DefaultTheme[mode].timerIcon.color,
      size: customStyle.timerIcon?.size || DefaultStyle.timerIcon.size,
    },
    notificationIcon: {
      size:
        customStyle?.notificationIcon?.size ||
        DefaultStyle?.notificationIcon?.size,
    },
    deleteIcon: {
      color: theme.colors?.deleteIcon || DefaultTheme[mode].colors.deleteIcon,
      size: customStyle.deleteIcon?.size || DefaultStyle.deleteIcon.size,
    },
    loadMoreButton: {
      color:
        theme?.loadMoreButton?.color ||
        theme.colors?.primaryColor ||
        DefaultTheme[mode]?.loadMoreButton?.color,
      backgroundColor:
        theme?.loadMoreButton?.background ||
        DefaultTheme[mode]?.loadMoreButton?.background,
      fontSize:
        customStyle?.loadMoreButton?.fontSize ||
        DefaultStyle?.loadMoreButton?.fontSize,
      fontWeight:
        customStyle.loadMoreButton?.fontWeight ||
        DefaultStyle.loadMoreButton.fontWeight,
    },
    loader: {
      backgroundImage: DefaultTheme[mode].loader.backgroundImage,
      borderColor:
        theme.colors?.primaryColor || DefaultTheme[mode].colors.primaryColor,
    },
    badgeStyle: {
      borderRadius: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      minWidth: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      height: customStyle.badgeStyle?.size || defaultBadgeStyle.size,
      backgroundColor: theme.badgeStyle?.color || defaultBadgeStyle.color,
      top:  customStyle?.badgeStyle?.top ? `${customStyle.badgeStyle.top}px` : defaultBadgeStyle.top,
      right:  customStyle?.badgeStyle?.right ? `${customStyle.badgeStyle.right}px` : defaultBadgeStyle.right,    
    },
    badgeTextStyle: {
      color: theme.badgeStyle?.textColor || defaultBadgeStyle.textColor,
      fontSize: customStyle.badgeStyle?.textSize || defaultBadgeStyle.textSize,
    },
    infiniteLoader: {
      border: `3px solid ${
        theme.colors?.infiniteLoader ||
        theme.colors?.primaryColor ||
        DefaultTheme[mode].colors.primaryColor
      }`,
    },
  };
};

export const isEmptyArray = (array: Array<NotificationDataType> = []) => {
  return array && array?.length === 0;
};
export const mergeArrays = (
  array1: Array<NotificationDataType> = [],
  array2: Array<NotificationDataType> = []
) => {
  if (array1 && array2) return [...array1, ...array2];
  if (array1) return array1;

  return array2;
};

export const isValidResponse = (
  response: NotificationsApiResponse | ActionResponse
): boolean => !!response?.data;

export const filterDataProperty = (
  response: NotificationsApiResponse
): NotificationDataType[] | null => {
  if (!response.data) return null;

  return response.data;
};

export const generateFilterParams = (
  data: NotificationDataType[],
  fromStart: boolean,
  itemsPerPage: number
): FetchParams => {
  let params: FetchParams = { size: itemsPerPage, sort: "createdAt" };

  if (data.length > 0)
    if (fromStart) params = { ...params, start: data[0].createdAt };
    else params = { ...params, end: data[data.length - 1].createdAt };

  return params;
};

export const calculateModalPosition = (
  iconRef: RefObject<HTMLDivElement>,
  window: Window,
  containerWidth: DimensionValue
) => {
  if (iconRef.current) {
    const iconRect = iconRef.current.getBoundingClientRect();
    const screenWidth = window.innerWidth;
    const spaceRight = screenWidth - iconRect.x;
    const spaceLeft = iconRect.x;
    let modalWidth = calculateModalWidth(containerWidth);

    const centerPosition =
      Math.min(spaceLeft, spaceRight) + Math.abs(spaceLeft - spaceRight) / 2;

    if (window.innerWidth <= modalWidth) modalWidth = window.innerWidth - 40;

    if (spaceRight > modalWidth) {
      return {
        left: `0px`,
      };
    } else if (spaceLeft > modalWidth) {
      const rightPosition = spaceRight < modalWidth ? "30px" : null;

      return {
        ...(rightPosition && { right: rightPosition }),
      };
    } else if (
      spaceLeft < modalWidth &&
      spaceRight < modalWidth &&
      spaceLeft > spaceRight
    ) {
      return { right: "30px" };
    } else {
      return { left: `-${centerPosition - 40}px` };
    }
  }
};

export const calculateModalWidth = (containerWidth: DimensionValue): number => {
  let modalWidth = 500;

  if (typeof containerWidth === "string")
    modalWidth = parseInt(containerWidth.slice(0, -2)) + 40;
  else if (typeof containerWidth === "number") modalWidth = containerWidth + 40;

  return modalWidth;
};

export const hexToRgba = (hex: string, alpha: number) => {
  const [r, g, b] = hex.match(/\w\w/g)?.map((x) => parseInt(x, 16)) ?? [];

  return `rgba(${r},${g},${b},${alpha})`;
};

export const debounce = <F extends (...args: unknown[]) => void>(
  func: F,
  delay: number
) => {
  let timerId: ReturnType<typeof setTimeout>;

  return (...args: Parameters<F>): void => {
    clearTimeout(timerId);
    timerId = setTimeout(() => {
      func(...args);
    }, delay);
  };
};
