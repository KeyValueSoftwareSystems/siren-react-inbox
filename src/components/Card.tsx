import type { CSSProperties } from "react";
import React, { type FC } from "react";

import CloseIcon from "./CloseIcon";
import TimerIcon from "./TimerIcon";
import DefaultAvatar from "../assets/defaultAvatar.svg";
import type { NotificationCardProps } from "../types";
import { generateElapsedTimeText } from "../utils/commonUtils";
import "../styles/card.css";

/**
 * Card component represents an individual notification card in the notification list.
 *
 * @component
 * @example
 * const notification = {
 *   id: "1",
 *   message: {
 *     header: "New Message",
 *     body: "You have a new message.",
 *     avatar: { imageUrl: "https://example.com/avatar.png" }
 *   },
 *   createdAt: "2024-02-24T12:00:00Z"
 * };
 *
 * <Card
 *   notification={notification}
 *   cardProps={{ hideAvatar: false }}
 *   styles={customStyles}
 *   onNotificationCardClick={(notification) => console.log('Notification clicked', notification)}
 *   deleteNotificationById={(id) => console.log('Notification deleted', id)}
 * />
 *
 * @param {NotificationCardProps} props - The properties passed to the Card component.
 * @param {Object} props.notification - The notification data to display in the card.
 * @param {Object} [props.cardProps] - Optional properties to customize the appearance of the card.
 * @param {Object} props.styles - Custom styles applied to the card and its elements.
 * @param {Function} [props.onNotificationCardClick] - Callback function executed when the card is clicked.
 * @param {Function} [props.deleteNotificationById] - Callback function executed when the delete action is triggered.
 * @returns {ReactElement} The rendered Card component.
 */

const Card: FC<NotificationCardProps> = ({
  notification,
  cardProps,
  styles,
  onNotificationCardClick,
  deleteNotificationById,
}) => {
  const { id, createdAt, message, isRead } = notification;
  const { avatar, header, subHeader, body } = message;

  const onDelete = (event: React.MouseEvent) => {
    deleteNotificationById(id);
    event.stopPropagation();
  };

  const cardContainerStyle: CSSProperties = isRead
    ? {
      ...styles.defaultCardContainer,
      borderLeft: "4px transparent solid",
    }
    : {
      ...styles.defaultCardContainer,
      borderLeft: `4px ${styles.activeCardMarker.border} solid`,
      backgroundColor: styles.activeCardMarker.backgroundColor,
    };

  return (
    <div
      style={cardContainerStyle}
      className={`${
        cardProps?.hideAvatar
          ? "siren-sdk-hide-avatar-card-container"
          : "siren-sdk-card-container"
      }`}
      onClick={() =>
        onNotificationCardClick && onNotificationCardClick(notification)
      }
      data-testid={`card-${notification.id}`}
    >
      {!cardProps?.hideAvatar && (
        <img
          src={avatar?.imageUrl || DefaultAvatar}
          alt=""
          style={styles.cardIconRound}
        />
      )}
      <div className="siren-sdk-card-content-wrapper">
        <div style={styles.cardTitle} className="siren-sdk-card-text-break">
          {header}
        </div>
        <div
          style={styles.cardDescription}
          className="siren-sdk-card-text-break"
        >
          {subHeader}
        </div>
        <div
          style={styles.cardDescription}
          className="siren-sdk-card-text-break siren-sdk-card-msg-body"
        >
          {body}
        </div>
        <div className="siren-sdk-card-date-container">
          <TimerIcon
            color={styles.timerIcon.color}
            size={styles.timerIcon.size}
          />
          <div style={styles.dateStyle}>
            {generateElapsedTimeText(createdAt)}
          </div>
        </div>
      </div>
      <div
        data-testid={`delete-${notification.id}`}
        className="siren-sdk-delete-button"
        onClick={onDelete}
      >
        <CloseIcon
          color={styles?.deleteIcon.color}
          size={styles.deleteIcon.size}
        />
      </div>
    </div>
  );
};

export default Card;
