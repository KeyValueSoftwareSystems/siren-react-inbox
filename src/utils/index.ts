import * as CommonUtils from './commonUtils';
import * as Constants from './constants';
import DefaultStyle from './defaultStyles'
import DefaultTheme from './defaultTheme';
import useSiren from './sirenHook';

export {
  Constants,
  DefaultTheme,
  useSiren,
  CommonUtils,
  DefaultStyle
};