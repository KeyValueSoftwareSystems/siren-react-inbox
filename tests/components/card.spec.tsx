import React from "react";

import { act, fireEvent, render } from "@testing-library/react";

import Card from "../../src/components/Card";
import { applyTheme } from "../../src/utils/commonUtils";

const mockNotification = {
  id: "1",
  createdAt: "",
  message: {
    header: "New Message",
    body: "You have a new message.",
    channel: "watsapp",
    subHeader: "New post",
    actionUrl: "",
    additionalData: "",
    avatar: {
      imageUrl: "",
      actionUrl: "",
    },
  },
  requestId: "123",
  isRead: false,
};

const mockDeleteNotificationById = jest.fn();
const mockOnNotificationCardClick = jest.fn();

// Mock the CSS file to avoid Jest error
jest.mock("../../src/styles/card.css", () => ({}));

const style = applyTheme();

test("matches snapshot", () => {
  const { asFragment } = render(
    <Card
      notification={mockNotification}
      deleteNotificationById={mockDeleteNotificationById}
      onNotificationCardClick={mockOnNotificationCardClick}
      cardProps={{}}
      styles={style}
      darkMode={false}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});
test("renders notification card with basic content", () => {
  const screen = render(
    <Card
      notification={mockNotification}
      deleteNotificationById={mockDeleteNotificationById}
      onNotificationCardClick={mockOnNotificationCardClick}
      cardProps={{}}
      styles={style}
      darkMode={false}
    />
  );

  expect(screen.getByTestId("card-1")).toBeDefined();
});

test("triggers delete notification callback on delete button click", () => {

  jest.useFakeTimers();

  const screen = render(
    <Card
      notification={mockNotification}
      deleteNotificationById={mockDeleteNotificationById}
      onNotificationCardClick={mockOnNotificationCardClick}
      cardProps={{}}
      styles={style}
      darkMode={false}
    />
  );

  const deleteButton = screen.getByTestId("delete-1");

  fireEvent.click(deleteButton);

  act(() => {
    jest.advanceTimersByTime(200);
  });

  expect(mockDeleteNotificationById).toHaveBeenCalledWith("1");

  jest.useRealTimers();
});

test("triggers notification card click callback on card click", () => {
  const screen = render(
    <Card
      notification={mockNotification}
      deleteNotificationById={mockDeleteNotificationById}
      onNotificationCardClick={mockOnNotificationCardClick}
      cardProps={{ hideAvatar: true }}
      styles={style}
      darkMode={false}
    />
  );

  const card = screen.getByTestId("card-1");

  fireEvent.click(card);
  expect(mockOnNotificationCardClick).toHaveBeenCalledWith(mockNotification);
});

test("renders header, subheader, and body", () => {
  const screen = render(
    <Card
      notification={mockNotification}
      deleteNotificationById={mockDeleteNotificationById}
      onNotificationCardClick={mockOnNotificationCardClick}
      cardProps={{}}
      styles={style}
      darkMode={false}
    />
  );

  const header = screen.getByText(mockNotification.message.header);
  const subHeader = screen.getByText(mockNotification.message.subHeader);
  const body = screen.getByText(mockNotification.message.body);

  expect(header).toBeTruthy();
  expect(subHeader).toBeTruthy();
  expect(body).toBeTruthy();
});

test("does not render avatar if hideAvatar is true", () => {
  const screen = render(
    <Card
      notification={mockNotification}
      deleteNotificationById={mockDeleteNotificationById}
      onNotificationCardClick={mockOnNotificationCardClick}
      cardProps={{ hideAvatar: true }}
      styles={style}
      darkMode={false}
    />
  );

  const avatarContainer = screen.queryByTestId("avatar-container");

  expect(avatarContainer).toBeNull();
});
