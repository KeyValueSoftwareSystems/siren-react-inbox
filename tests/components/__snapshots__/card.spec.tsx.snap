// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="siren-notification-card-1"
    class="siren-sdk-card-container siren-sdk-card-common-container"
    data-testid="card-1"
    style="background-color: rgb(46, 45, 48); padding: 10px; border-bottom: 0.5px solid; border-color: #344054; border-left: 4px solid #F56630;"
  >
    <div
      aria-label="siren-notification-avatar-1"
      style="width: 40px; height: 40px; border-radius: 20px; overflow: hidden; background-color: rgb(52, 64, 84); background-size: cover; background-position: center;"
    />
    <div
      class="siren-sdk-card-content-wrapper"
    >
      <div
        class="siren-sdk-card-text-break"
        style="color: rgb(255, 255, 255); font-size: 16px; font-weight: 600;"
      >
        New Message
      </div>
      <div
        class="siren-sdk-card-text-break"
        style="color: rgb(255, 255, 255); font-size: 14px; font-weight: 400;"
      >
        New post
      </div>
      <div
        class="siren-sdk-card-text-break siren-sdk-card-msg-body"
        style="color: rgb(255, 255, 255); font-size: 14px; font-weight: 400;"
      >
        You have a new message.
      </div>
      <div
        class="siren-sdk-card-date-container"
      >
        <svg
          fill="none"
          height="12"
          viewBox="0 0 14 13"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.99723 0.0807266C3.45907 0.0807266 0.580566 2.95923 0.580566 6.49739C0.580566 10.0356 3.45907 12.9141 6.99723 12.9141C10.5354 12.9141 13.4139 10.0356 13.4139 6.49739C13.4139 2.95923 10.5354 0.0807266 6.99723 0.0807266ZM6.99723 11.7474C4.1025 11.7474 1.74723 9.39213 1.74723 6.49739C1.74723 3.60266 4.1025 1.24739 6.99723 1.24739C9.89197 1.24739 12.2472 3.60266 12.2472 6.49739C12.2472 9.39213 9.89197 11.7474 6.99723 11.7474Z"
            fill="#98A2B3"
          />
          <path
            d="M10.1553 7.90104C10.0588 8.11837 9.84542 8.2474 9.62211 8.24768C9.54293 8.24768 9.46261 8.23145 9.38541 8.19698L6.76041 7.03031C6.54993 6.93689 6.41406 6.72782 6.41406 6.4974V2.9974C6.41406 2.67525 6.67525 2.41406 6.9974 2.41406C7.31954 2.41406 7.58073 2.67525 7.58073 2.9974V6.11829L9.85937 7.13114C10.1536 7.26188 10.2863 7.60653 10.1553 7.90104Z"
            fill="#98A2B3"
          />
        </svg>
        <div
          style="color: rgb(152, 162, 179); font-size: 12px; line-height: 16px;"
        >
          NaN years ago
        </div>
      </div>
    </div>
    <div
      aria-label="siren-notification-delete-1"
      class="siren-sdk-delete-button"
      data-testid="delete-1"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 20 20"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 15L15 5M5 5L15 15"
          stroke="#D0D5DD"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
